File: src/main/scala/com/vc/medium/L684.scala
package com.vc.medium

object L684 {
  def findRedundantConnection(edges: Array[Array[Int]]): Array[Int] = {
    val pArr = new Array[Int](edges.length + 1)
    pArr.indices.foreach(i => {
      pArr(i) = i
    })
    def union(x: Int, y: Int): Unit = pArr(x) = y
    def find(x: Int): Int = if(pArr(x) == x) x else find(pArr(x))
    edges.foreach(e => {
      val x = find(e(0))
      val y = find(e(1))
      if(x == y) return e
      union(x, y)
    })
    Array.empty[Int]
  }
}

File: src/main/java/com/vc/medium/L684.java
package com.vc.medium;

class L684 {
    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;

        int[] parent = new int[n + 1];
        for(int i = 0; i < parent.length; i++) parent[i] = i;
        for(int[] edge: edges) {
            int from = edge[0];
            int to = edge[1];

            int fromParent = find(from, parent);
            int toParent = find(to, parent);

            if(fromParent == toParent) return edge;
            parent[fromParent] = toParent;
        }
        return new int[0];
    }

    private int find(int x, int[] parent) {
        if(x == parent[x]) return x;
        else return find(parent[x], parent);
    }
}
